File: Container/MXDocument.m
//
//  MXDocument.m
//  Lines
//
//  Created by Paolo Bosetti on 26/09/14.
//  Copyright (c) 2014 University of Trento. All rights reserved.
//

#import "MXDocument.h"

@implementation MXDocument

+ (NSString *)humanName
{
  return @"Generic file";
}

+ (NSArray *)extensions
{
  return @[@""];
}

+ (NSString *)UTI
{
  return @"public.plain-text";
}

+ (NSMutableArray *)filterFileNames:(NSArray *)list atPath:(NSString *)path
{
  NSPredicate    *filter;
  NSMutableArray *shortList = [NSMutableArray arrayWithCapacity:list.count];
  NSMutableArray *result = [NSMutableArray arrayWithCapacity:list.count];
  
  for (NSString *ext in [self extensions]) {
    filter = [NSPredicate predicateWithFormat:@"self ENDSWITH %@", ext];
    [shortList removeAllObjects];
    [shortList addObjectsFromArray:[list filteredArrayUsingPredicate:filter]];
    for (NSString *fileName in shortList) {
      [result addObject:[[self alloc] initWithFileName:fileName]];
      [[result lastObject] setFileExt:ext];
      [[result lastObject] setFolderPath:path];
    }
  }
  
  return result;
}


- (id)init
{
  self = [super init];
  if (self) {
    self.folderPath = [NSHomeDirectory() stringByAppendingPathComponent:@"Documents"];
    _humanName = [self.class humanName];
  }
  return self;
}

- (instancetype)initWithFileName:(NSString *)name
{
  self = [self init];
  if (self) {
    self.fileName = name;
  }
  return self;
}

- (NSString *)filePath
{
  return [self.folderPath stringByAppendingPathComponent:_fileName];
}

- (NSURL *)fileURL
{
  return [NSURL fileURLWithPath:[self filePath]];
}



- (BOOL) delete
{
  return NO;
}

- (BOOL) renameTo:(NSString *)newName
{
  return NO;
}


- (void)renderInView:(UIView *)view
{
  [view addSubview:[[UIView alloc] init]];
}

- (NSString *)fileDescription
{
  NSDictionary *attr = [[NSFileManager defaultManager] attributesOfItemAtPath:self.filePath error:nil];
  return [NSString stringWithFormat:@"%@ (%lu kb)", self.fileName, [attr[NSFileSize] integerValue]  / 1024];
}

@end

File: Container/MXDocument.swift
//
//  Document.swift
//  Lines
//
//  Created by Paolo Bosetti on 02/10/14.
//  Copyright (c) 2014 University of Trento. All rights reserved.
//

import Foundation
import UIKit

class MXDocument: NSObject {
  var folderPath: String
  var fileName: String
  
  var fileExt: String {
    get { return fileName.pathExtension }
  }
  
  var filePath: String {
    get { return "\(folderPath)/\(fileName)"}
  }
  
  var fileURL: NSURL {
    get { return NSURL(fileURLWithPath: self.filePath) }
  }
  
  var fileExists: Bool {
    get { return NSFileManager.defaultManager().fileExistsAtPath(self.filePath) }
  }
  
  var fileAttr: [NSObject: AnyObject] {
    get {
      if self.fileExists {
        var attr: [NSObject: AnyObject] = NSFileManager.defaultManager().attributesOfItemAtPath(self.filePath, error: nil)!
        return attr
      }
      else {
        return [:]
      }
    }
  }
  
  var fileDescription: String {
    get {
      if (self.fileExists) {
        return String(format: "%@ (%lu kb)", self.fileName, self.fileAttr[NSFileSize]!.integerValue / 1024)
      }
      else {
        return "File \(self.fileName) does not exist!"
      }
    }
  }
  
  class func humanName() -> String {
    return "Generic file"
  }
  
  class func UTI() -> String {
    return "public.plain-text"
  }
  
  class func extensions() -> [String] {
    return [""]
  }
  
  class func filterFileNames(list: [String], atPath: String) -> [MXDocument] {
    var result :[MXDocument] = []
    for ext in self.extensions() {
      for name in list {
        if name.pathExtension == ext {
          result.append(MXDocument(fileName: name))
        }
      }
    }
    return result
  }
  
  override init() {
    folderPath = NSHomeDirectory().stringByAppendingPathComponent("Documents")
    fileName = ""
  }
  
  convenience init(fileName name: String) {
    self.init()
    fileName = name
  }
  
  func inspect() {
    NSLog("%@", self.fileDescription)
  }
  
  
}